### PR TITLE
Rate limiter; issue with password locking fixed; faster signing of S3 items

### PR DESCRIPTION
**Pull request by UnknownJoe796**
See commit

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -15,6 +15,10 @@ buildscript {
     }
 }
 
+allprojects {
```
Comment (bjsvedin) [inline build.gradle.kts:18]: This seems to be a duplicate.

File: server-aws/src/main/kotlin/com/lightningkite/lightningserver/files/S3File.kt [hunk 2]
```diff
@@ -168,15 +171,101 @@ data class S3File(val system: S3FileSystem, val path: File) : FileObject {
...
     override val signedUrl: String
         get() = system.signedUrlDuration?.let { e ->
-            system.signer.presignGetObject {
-                it.signatureDuration(system.signedUrlDuration.toJavaDuration())
-                it.getObjectRequest {
-                    it.bucket(system.bucket)
-                    it.key(path.unixPath)
+            var stopwatch = TimeSource.Monotonic.markNow()
+            var timeIndex = 0
+            fun mark() {
```
Comment (bjsvedin) [inline server-aws/src/main/kotlin/com/lightningkite/lightningserver/files/S3File.kt:176]: Do we still need the benchmarking?

File: server-core/src/main/kotlin/com/lightningkite/lightningserver/db/ModelRestUpdatesWebsocket.kt [hunk 3]
```diff
@@ -131,13 +132,15 @@ class ModelRestUpdatesWebsocket<USER: HasId<*>?, T : HasId<ID>, ID : Comparable<
             }.filter { it.old != null || it.new != null }
 
             jobs.add(launch {
-                if (toSend.size > 20) {
+                val updates = CollectionUpdates(
+                    updates = toSend.mapNotNull { it.new }.toSet(),
+                    remove = toSend.mapNotNull { it.old.takeIf { _ -> it.new == null }?._id }.toSet()
+                )
+                val size = Serialization.json.encodeToString(CollectionUpdates.serializer(info.serialization.serializer, info.serialization.idSerializer), updates).length
+                if (size >= 24000) {
```
Comment (bjsvedin) [inline server-core/src/main/kotlin/com/lightningkite/lightningserver/db/ModelRestUpdatesWebsocket.kt:140]: This seems rather large if the last one was 20. Is this correct?

File: server-core/src/main/kotlin/com/lightningkite/lightningserver/http/Http.kt [hunk 4]
```diff
@@ -75,6 +77,8 @@ object Http {
 
     suspend fun execute(request: HttpRequest): HttpResponse {
         return endpoints[request.endpoint]?.let { handler ->
+            val authOrNull = request.authAny()
+            serverLogger.info("${request.endpoint} (${request.parts}) accessed by ${authOrNull} (${request.sourceIp})")
```
Comment (bjsvedin) [inline server-core/src/main/kotlin/com/lightningkite/lightningserver/http/Http.kt:81]: I suggest this go to debug not info.

File: server-core/src/main/kotlin/com/lightningkite/lightningserver/typed/BulkRequestEndpoint.kt [hunk 5]
```diff
@@ -32,6 +34,7 @@ fun ServerPath.bulkRequestEndpoint() = post.api(
                     val api = Http.endpoints[handler.endpoint] as? ApiEndpoint<HasId<*>?, TypedServerPath, Any?, Any?> ?: return@async entry.key to BulkResponse(
                         error = LSError(400, detail = "not-api", message = "Matched route is not an API endpoint", data = it.method + " " + it.path)
                     )
+                    serverLogger.info("${api.route.endpoint} (${handler.parts}) accessed by ${authOrNull} (${rawRequest?.sourceIp}) (via bulk)")
```
Comment (bjsvedin) [inline server-core/src/main/kotlin/com/lightningkite/lightningserver/typed/BulkRequestEndpoint.kt:37]: I suggest this go to debug not info.

File: server-redis/src/main/kotlin/com/lightningkite/lightningserver/cache/RedisCache.kt [hunk 6]
```diff
@@ -88,8 +93,14 @@ class RedisCache(val client: RedisClient) : Cache {
 //    }
 
     override suspend fun add(key: String, value: Int, timeToLive: Duration?) {
+        println("add $key $value $timeToLive")
```
Comment (bjsvedin) [inline server-redis/src/main/kotlin/com/lightningkite/lightningserver/cache/RedisCache.kt:96]: Please remove printlns

File: client/src/commonMain/kotlin/com/lightningkite/lightningserver/db/ModelCache3.kt [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+                    }
+                }
+            }
+    }
+
+    private fun individualItemUpdates(
+        values: List<T>,
+        it: List<ItemHolder>
+    ) {
+        println("individualItemUpdates: $values $it")
```
Comment (bjsvedin) [inline client/src/commonMain/kotlin/com/lightningkite/lightningserver/db/ModelCache3.kt:170]: Please remove println

**Review by bjsvedin — CHANGES_REQUESTED** [review]

File: server-aws/src/main/kotlin/com/lightningkite/lightningserver/files/S3File.kt (hunk 2 above)
Comment (UnknownJoe796) [reply server-aws/src/main/kotlin/com/lightningkite/lightningserver/files/S3File.kt:176]: Removing benchmarking

File: server-core/src/main/kotlin/com/lightningkite/lightningserver/db/ModelRestUpdatesWebsocket.kt (hunk 3 above)
Comment (UnknownJoe796) [reply server-core/src/main/kotlin/com/lightningkite/lightningserver/db/ModelRestUpdatesWebsocket.kt:140]: It's now measuring payload size, not item count

**Pull request closed without merging**